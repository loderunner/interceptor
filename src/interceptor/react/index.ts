export { Provider as InterceptProvider, useIntercepts } from './intercept'
