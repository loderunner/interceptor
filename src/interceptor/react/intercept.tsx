import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
