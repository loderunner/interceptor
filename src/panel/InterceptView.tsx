import * as React from 'react'
import { useCallback } from 'react'

import { Intercept } from '@/interceptor'
import { useIntercept } from '@/interceptor/hooks'

import { useSelection } from './selection'

interface Props {
  inter: Intercept
}

const InterceptView = ({ inter }: Props) => {
  const { selection, setSelection } = useSelection()
  const { intercept, setIntercept } = useIntercept(inter)

  const onToggleEnabled = useCallback(() => {
    const inter = { ...intercept, enabled: !intercept.enabled }
    setIntercept(inter)
    if (selection === intercept) {
      setSelection(inter)
    }
  }, [selection, intercept])

  return (
    <div className="max-w-5xl mx-24 mt-8 px-8 pt-2 grid grid-cols-1 gap-6">
      <label className="block">
        <span className="text-3xl font-bold">Intercept</span>
      </label>
      <label className="block">
        <input
          type="checkbox"
          className="mt-1"
          checked={intercept.enabled}
          onChange={onToggleEnabled}
        />
        <span className="ml-1 text-sm text-gray-700">Enabled</span>
      </label>
      <label className="block">
        <span className="text-sm text-gray-700">If URL matches</span>
        <input
          type="text"
          className="mt-1 block text-sm w-full"
          placeholder="example.com"
          value={intercept.pattern}
          onChange={() => {}}
        />
      </label>
    </div>
  )
}

export default InterceptView
