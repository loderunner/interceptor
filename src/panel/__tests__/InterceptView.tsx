import { cleanup, render } from '@testing-library/react'
import * as React from 'react'
import '@testing-library/jest-dom'

import InterceptView from '../InterceptView'

jest.mock('@/interceptor/hooks')

describe('[InterceptView]', () => {
  afterEach(() => {
    cleanup()
  })

  it('should match initial snapshot', () => {
    const { container } = render(<InterceptView interceptId="inter" />)
    expect(container).toMatchSnapshot()
  })
})
