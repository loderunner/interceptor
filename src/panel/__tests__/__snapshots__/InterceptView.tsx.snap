// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[InterceptView] should match initial snapshot 1`] = `
<div>
  <div
    class="max-w-5xl mx-24 mt-8 px-8 pt-2 grid grid-cols-1 gap-6"
  >
    <label
      class="block"
    >
      <span
        class="text-3xl font-bold"
      >
        Intercept
      </span>
    </label>
    <label
      class="mt-1 inline-flex items-center"
    >
      <input
        checked=""
        type="checkbox"
      />
      <span
        class="ml-1 text-sm text-gray-700"
      >
        Enabled
      </span>
    </label>
    <label
      class="mt-1 block"
    >
      <span
        class="text-sm text-gray-700"
      >
        If URL matches
      </span>
      <div
        class="flex items-center space-x-2"
      >
        <input
          class="block mt-1 text-sm flex-auto"
          placeholder="example.com"
          type="text"
          value="example.com"
        />
        <label
          class="inline-flex items-center"
        >
          <input
            type="checkbox"
          />
          <span
            class="ml-1 text-sm text-gray-700"
          >
            Regexp
          </span>
        </label>
      </div>
    </label>
    <label
      class="mt-1 inline-flex items-center"
    >
      <input
        type="checkbox"
      />
      <span
        class="ml-1 text-sm text-gray-700"
      >
        Intercept response
      </span>
    </label>
  </div>
</div>
`;
